In [344]:
import pandas as pd
import numpy as np

In [345]:
df = pd.read_csv('/content/sample_data/california_housing_train.csv')

## Understanding the new dataset

In [346]:
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


In [347]:
print(f"The number of rows in the dataset is {df.shape[0]}")
print(f"The number of columns in the dataset is {df.shape[1]}")

The number of rows in the dataset is 17000
The number of columns in the dataset is 9


### The next code cell basically gives information about the dataframe. It contains both dependent and independent variables.

In [348]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17000 entries, 0 to 16999
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           17000 non-null  float64
 1   latitude            17000 non-null  float64
 2   housing_median_age  17000 non-null  float64
 3   total_rooms         17000 non-null  float64
 4   total_bedrooms      17000 non-null  float64
 5   population          17000 non-null  float64
 6   households          17000 non-null  float64
 7   median_income       17000 non-null  float64
 8   median_house_value  17000 non-null  float64
dtypes: float64(9)
memory usage: 1.2 MB


This is a summary of a DataFrame in pandas, specifically providing details about the index and columns within the DataFrame. Here’s what each part means:

1. **RangeIndex: 17000 entries, 0 to 16999**
   - The DataFrame has **17,000 rows**, with the row indices ranging from **0 to 16,999**.
   - **RangeIndex** is the default index assigned by pandas, which is simply a sequence of integers. The index helps in identifying the rows and accessing specific rows if necessary.

2. **Data columns (total 9 columns)**
   - This means that the DataFrame contains **9 columns**.

3. **Column Information:**
   - For each column, pandas provides:
     - The column number (starting from 0).
     - The **name** of the column.
     - The **Non-Null Count**, which is the number of entries in each column that are not null (missing values). In this case, each column has 17,000 non-null entries, indicating there are no missing values in any of the columns.
     - The **Dtype** of the column, which refers to the data type. All columns in this DataFrame are of type `float64`, which indicates that the columns contain floating-point numbers.

4. **Columns in Detail:**
   - **longitude**: Contains the longitude coordinates (a `float64` value).
   - **latitude**: Contains the latitude coordinates.
   - **housing_median_age**: Median age of houses in that region.
   - **total_rooms**: Total number of rooms in the area.
   - **total_bedrooms**: Total number of bedrooms.
   - **population**: The population of that area.
   - **households**: Number of households in the area.
   - **median_income**: The median income of households in the area.
   - **median_house_value**: Median value of houses.

5. **Dtypes: float64(9)**
   - All 9 columns have the `float64` data type. This indicates that each value in these columns is a 64-bit floating-point number, suitable for continuous numeric data.

6. **Memory usage: 1.2 MB**
   - The estimated memory usage of the DataFrame is **1.2 MB**, which represents how much memory the DataFrame occupies.

This summary helps you quickly understand the structure of your data: it tells you the shape of the DataFrame (17,000 rows and 9 columns), the type of data in each column, and whether there are any missing values.

### Distribution of these features by plotting their histograms.

### Method to implement Scaling

In [349]:
def standard_scaler(data):
  """
    Args:
      data: A pandas DataFrame.

    Returns:
      A new DataFrame with standardized data.
  """
  # mean is a Series of mean values for the column
  mean = data.mean()
  # std is a series of standard values for all the column
  std = data.std()
  # scaled_data is the dataframe that has been standarized
  scaled_data = (data - mean) / std
  return scaled_data

In [350]:
# We need to drop the target column before standarizing the dataset
target_column = 'median_house_value'
dataset_beforescaling = df.drop(target_column, axis = 1)
data_scaled = standard_scaler(dataset_beforescaling)
data_scaled.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income
0,2.619288,-0.671500,-1.079639,1.361655,1.764152,-0.361173,-0.075996,-1.252506
1,2.539494,-0.573248,-0.761850,2.296540,3.230346,-0.261858,-0.099401,-1.081451
2,2.494610,-0.905436,-0.920744,-0.882436,-0.866931,-0.955326,-0.999223,-1.170071
3,2.489623,-0.928830,-1.159087,-0.524171,-0.480216,-0.796769,-0.715753,-0.362590
4,2.489623,-0.961581,-0.682402,-0.545731,-0.506313,-0.701809,-0.622130,-1.026424


In [351]:
X = data_scaled.values #Converting the data frame to array values
Y = df[target_column].values # Converting the target column in the dataframe to array

### Linear Regression from scratch

In [352]:
class LinearRegressionScratch():
  def __init__( self, learning_rate, iterations ):
    self.learning_rate = learning_rate
    self.iterations = iterations

  def predict( self, X ):
    return X.dot( self.W ) + self.b

  def update_weights(self):
    Y_pred = self.predict( self.X )
    dW = - ( 2 * ( self.X.T ).dot( self.Y - Y_pred )  ) / self.m
    db = - 2 * np.sum( self.Y - Y_pred ) / self.m
    self.W = self.W - self.learning_rate * dW
    self.b = self.b - self.learning_rate * db
    return self

  def fit( self, X, Y ):
    self.m, self.n = X.shape
    self.W = np.random.rand(self.n)
    self.b = 0
    self.X = X
    self.Y = Y
    for i in range(self.iterations):
      if i%10==0:
        print(f"At {i}th iteration we have value of: \n W as :{self.W} and \n b as {self.b}\n")
      self.update_weights()
    return self

In [353]:
model = LinearRegressionScratch(iterations = 300, learning_rate = 0.02)
model.fit(X, Y)

At 0th iteration we have value of: 
 W as :[0.5597281  0.35424577 0.53436389 0.11163755 0.81297465 0.92707538
 0.85663147 0.4872479 ] and 
 b as 0

At 10th iteration we have value of: 
 W as :[-2523.25870358 -5592.06261601  5035.23905642  4174.76583337
  1271.83894032 -2105.56124392  1857.77313104 26754.4295336 ] and 
 b as 69480.50034989217

At 20th iteration we have value of: 
 W as :[-5379.72291332 -9593.14446197  9506.12111142  6061.86245368
  2054.45841654 -4629.88443202  3013.85346496 44399.16065182] and 
 b as 115673.4045475194

At 30th iteration we have value of: 
 W as :[ -8136.81696338 -12784.38675094  13084.18660332   6935.4484457
   2982.27994188  -6925.51482941   4170.0110901   56088.92370071] and 
 b as 146383.95480933078

At 40th iteration we have value of: 
 W as :[-10694.90483854 -15510.13758048  15833.2309217    7220.88394945
   4055.57738007  -8974.2185575    5372.03439898  63854.7426444 ] and 
 b as 166801.33089264034

At 50th iteration we have value of: 
 W as :[-1

### Linear Regression model to test

In [354]:
df_test = pd.read_csv('/content/sample_data/california_housing_test.csv')

In [355]:
target_column = 'median_house_value'
dataset_beforescaling = df_test.drop(target_column, axis=1)
data_scaled = standard_scaler(dataset_beforescaling)

In [356]:
X = data_scaled.values
Y = df_test['median_house_value'].values

In [357]:
predictions = model.predict(X)
print("Predictions:\n", predictions)

Predictions:
 [346549.58563287 220415.4906721  285449.82947891 ...  90942.09603831
 172470.10036447 455882.06749027]


In [358]:
import numpy as np
from sklearn.metrics import r2_score

def calculate_accuracy(model, X, Y):
    predictions = model.predict(X)
    r2_scores = r2_score(Y, predictions, multioutput='raw_values')
    for idx, score in enumerate(r2_scores):
        print(f"R² score for output variable {idx+1}: {score:.4f}")
        return r2_scores

In [359]:
calculate_accuracy(model, X, Y)

R² score for output variable 1: 0.6031


array([0.60314849])